src/main/java/io/vertx/core/impl/BlockedThreadChecker.java

```diff
@@ -52,24 +83,28 @@ public void run() {
             long dur = now - execStart;
             final long timeLimit = entry.getValue().maxExecTime();
             TimeUnit maxExecTimeUnit = entry.getValue().maxExecTimeUnit();
-            long val = maxExecTimeUnit.convert(dur, TimeUnit.NANOSECONDS);
-            if (execStart != 0 && val >= timeLimit) {
-              final String message = "Thread " + entry.getKey() + " has been blocked for " + (dur / 1_000_000) + " ms, time limit is " + TimeUnit.MILLISECONDS.convert(timeLimit, maxExecTimeUnit) + " ms";
-              if (warningExceptionTimeUnit.convert(dur, TimeUnit.NANOSECONDS) <= warningExceptionTime) {
-                log.warn(message);
-              } else {
-                VertxException stackTrace = new VertxException("Thread blocked");
-                stackTrace.setStackTrace(entry.getKey().getStackTrace());
-                log.warn(message, stackTrace);
-              }
+            long maxExecTimeInNanos = TimeUnit.NANOSECONDS.convert(timeLimit, maxExecTimeUnit);
+            long warningExceptionTimeInNanos = TimeUnit.NANOSECONDS.convert(warningExceptionTime, warningExceptionTimeUnit);
+            BlockedThreadEvent bts = new BlockedThreadEvent(entry.getKey(), dur, maxExecTimeInNanos, warningExceptionTimeInNanos);
+            if (execStart != 0 && dur >= maxExecTimeInNanos) {
+              blockedThreadHandler.handle(bts);
             }
           }
         }
       }
     }, intervalUnit.toMillis(interval), intervalUnit.toMillis(interval));
   }
 
-  synchronized void registerThread(Thread thread, Task checked) {
+  /**
+   * Specify the handler to run when it is determined a thread has been blocked for longer than allowed.
+   *
```

Comment:
we should specify that the handler will be call on the checker thread and not an event-loop